File: pkg-py/src/plotomics/upset.py
"""UpSet set-intersection widget."""

from __future__ import annotations

from typing import Any

import numpy as np

from ._base import STATIC, PlotomicsWidget, _column, _to_float32, pack_columns


class Upset(PlotomicsWidget):
    """Set intersections as a bar chart over a membership matrix.

    Venn diagrams stop being readable at four sets and stop being drawable at
    five. UpSet replaces the areas with an explicit matrix, so it scales to
    dozens of sets and stays exact.

    Intersections are **exclusive**: a column counts the elements in precisely
    that combination of sets and no others. That is what makes the columns sum
    to the union rather than double-counting, and it is why a small ``A + B``
    bar next to large ``A`` and ``B`` bars is evidence of mutual exclusivity
    rather than an artefact.

    Parameters
    ----------
    data:
        A pandas ``DataFrame`` or mapping with a numeric ``size`` column, one
        row per intersection, in the order to draw them.
    sets:
        Set names, top to bottom.
    membership:
        Boolean or 0/1 array shaped (intersections, sets).
    set_sizes:
        Per-set totals for the left-hand bars.
    total:
        Universe size, shown in the corner.
    bar_fraction:
        Fraction of the height given to the intersection bars.
    show_set_sizes:
        Draw the per-set total bars.
    dot_radius:
        Matrix dot radius in pixels.
    bar_color, empty_dot_color:
        Fills for the bars and filled dots, and for dots not in the
        intersection. ``None`` uses the theme.
    y_label:
        Axis label for the intersection bars.
    theme:
        Optional theme overrides forwarded to the JS renderer.
    height:
        Initial widget height in CSS pixels.

    Examples
    --------
    >>> import numpy as np, pandas as pd
    >>> m = np.array([[1, 0], [0, 1], [1, 1]])
    >>> Upset(pd.DataFrame({"size": [40, 25, 12]}), sets=["A", "B"],
    ...       membership=m)  # doctest: +SKIP
    """

    _esm = STATIC / "upset.js"

    def __init__(
        self,
        data: Any,
        *,
        sets: list[str],
        membership: Any,
        set_sizes: Any = None,
        total: float | None = None,
        bar_fraction: float = 0.55,
        show_set_sizes: bool = True,
        dot_radius: float = 5.0,
        bar_color: str | None = None,
        empty_dot_color: str | None = None,
        y_label: str = "intersection size",
        theme: dict | None = None,
        height: int = 520,
        **kwargs: Any,
    ) -> None:
        col = _column(data, "size")
        if col is None:
            raise ValueError("`data` must provide a `size` column.")
        size = _to_float32(col, "size")
        if size.size == 0:
            raise ValueError("`data` must contain at least one row.")

        names = [str(s) for s in sets]
        if not names:
            raise ValueError("`sets` must name at least one set.")

        m = np.atleast_2d(np.asarray(membership)).astype(int)
        if m.shape[0] != size.size:
            raise ValueError("`membership` must have one row per intersection.")
        if m.shape[1] != len(names):
            raise ValueError("`membership` must have one column per set.")

        buffer, schema = pack_columns({"size": size})

        meta: dict[str, Any] = {
            "sets": names,
            # Row-major: the component indexes it as intersection * nSets + set.
            "membership": [int(v) for v in m.reshape(-1)],
        }
        if set_sizes is not None:
            sizes = [float(v) for v in np.asarray(set_sizes).reshape(-1)]
            if len(sizes) != len(names):
                raise ValueError("`set_sizes` must have one entry per set.")
            meta["setSizes"] = sizes
        if total is not None:
            meta["total"] = float(total)

        options: dict[str, Any] = {
            "barFraction": bar_fraction,
            "showSetSizes": show_set_sizes,
            "dotRadius": dot_radius,
            "yLabel": y_label,
        }
        if bar_color is not None:
            options["barColor"] = str(bar_color)
        if empty_dot_color is not None:
            options["emptyDotColor"] = str(empty_dot_color)
        if theme is not None:
            options["theme"] = theme

        super().__init__(
            buffer=buffer,
            schema=schema,
            data={"columns": {}, "meta": meta},
            options=options,
            _height=height,
            **kwargs,
        )

File: pkg-r/R/upset.R
#' UpSet plot of set intersections
#'
#' Set intersections as a bar chart over a membership matrix. Venn diagrams stop
#' being readable at four sets and stop being drawable at five; UpSet replaces
#' the areas with an explicit matrix, so it scales to dozens of sets and stays
#' exact.
#'
#' Intersections are **exclusive**: a column counts the elements in precisely
#' that combination of sets and no others. That is what makes the columns sum to
#' the union rather than double-counting, and it is why a small `A + B` bar next
#' to large `A` and `B` bars is evidence of mutual exclusivity rather than an
#' artefact. [upset_intersections()] computes them from a logical matrix.
#'
#' @param data A data frame with a numeric `size` column, one row per
#'   intersection, in the order to draw them.
#' @param sets Character vector of set names, top to bottom.
#' @param membership Logical or 0/1 matrix, intersections x sets, saying which
#'   sets each intersection belongs to.
#' @param set_sizes Numeric vector of per-set totals for the left-hand bars.
#' @param total Universe size, shown in the corner.
#' @param bar_fraction Fraction of the height given to the intersection bars.
#' @param show_set_sizes Draw the per-set total bars.
#' @param dot_radius Matrix dot radius in pixels.
#' @param bar_color,empty_dot_color Fills for the bars and filled dots, and for
#'   dots not in the intersection. `NULL` uses the theme.
#' @param y_label Axis label for the intersection bars.
#' @param theme Optional named list of theme overrides.
#' @param width,height Widget dimensions (any valid CSS size).
#' @param element_id Optional explicit DOM id.
#' @return An `htmlwidget` object.
#' @examples
#' m <- matrix(c(TRUE, FALSE, FALSE, TRUE, TRUE, TRUE), nrow = 3, byrow = TRUE)
#' upset(data.frame(size = c(40, 25, 12)), sets = c("A", "B"), membership = m)
#' @export
upset <- function(data,
                  sets,
                  membership,
                  set_sizes = NULL,
                  total = NULL,
                  bar_fraction = 0.55,
                  show_set_sizes = TRUE,
                  dot_radius = 5,
                  bar_color = NULL,
                  empty_dot_color = NULL,
                  y_label = "intersection size",
                  theme = NULL,
                  width = NULL,
                  height = NULL,
                  element_id = NULL) {
  if (!is.data.frame(data)) {
    stop("`data` must be a data frame.", call. = FALSE)
  }
  if (is.null(data$size)) {
    stop("`data` must contain a `size` column.", call. = FALSE)
  }
  bv_require_nonempty(nrow(data), "data")
  sets <- as.character(sets)
  if (length(sets) == 0L) {
    stop("`sets` must name at least one set.", call. = FALSE)
  }

  size <- bv_require_numeric(data$size, "size")
  bv_check_finite(size, "size")

  m <- as.matrix(membership)
  if (nrow(m) != nrow(data)) {
    stop("`membership` must have one row per intersection.", call. = FALSE)
  }
  if (ncol(m) != length(sets)) {
    stop("`membership` must have one column per set.", call. = FALSE)
  }
  storage.mode(m) <- "integer"

  meta <- list(
    sets = I(sets),
    # Row-major: the component indexes it as intersection * nSets + set.
    membership = I(as.integer(t(m)))
  )
  if (!is.null(set_sizes)) {
    if (length(set_sizes) != length(sets)) {
      stop("`set_sizes` must have one entry per set.", call. = FALSE)
    }
    meta$setSizes <- I(as.numeric(set_sizes))
  }
  if (!is.null(total)) meta$total <- as.numeric(total)[1]

  options <- list(
    barFraction = bar_fraction,
    showSetSizes = show_set_sizes,
    dotRadius = dot_radius,
    yLabel = y_label
  )
  if (!is.null(bar_color)) options$barColor <- as.character(bar_color)[1]
  if (!is.null(empty_dot_color)) {
    options$emptyDotColor <- as.character(empty_dot_color)[1]
  }
  if (!is.null(theme)) options$theme <- theme

  plotomics_widget(
    "upset", list(size = size),
    meta = meta,
    options = options,
    width = width, height = height, element_id = element_id
  )
}

#' Exclusive set intersections from a membership matrix
#'
#' Counts the elements belonging to precisely each observed combination of sets.
#' Elements in no set are excluded, since they have no column to sit in.
#'
#' @param m A logical matrix, elements x sets, with set names as column names.
#' @param max_n Keep only the `max_n` largest intersections. `NULL` keeps all.
#' @return A list with `size` (integer vector), `membership` (logical matrix,
#'   intersections x sets), `sets`, `set_sizes` and `total`.
#' @examples
#' m <- cbind(A = c(TRUE, TRUE, FALSE), B = c(TRUE, FALSE, TRUE))
#' upset_intersections(m)
#' @export
upset_intersections <- function(m, max_n = NULL) {
  m <- as.matrix(m)
  if (is.null(colnames(m))) {
    stop("`m` must have column names naming the sets.", call. = FALSE)
  }
  storage.mode(m) <- "logical"
  m[is.na(m)] <- FALSE

  keys <- apply(m, 1L, function(r) paste0(as.integer(r), collapse = ""))
  empty <- strrep("0", ncol(m))
  keys <- keys[keys != empty]
  if (length(keys) == 0L) {
    return(list(size = integer(0),
                membership = matrix(FALSE, 0, ncol(m),
                                    dimnames = list(NULL, colnames(m))),
                sets = colnames(m), set_sizes = colSums(m), total = nrow(m)))
  }
  tab <- sort(table(keys), decreasing = TRUE)
  if (!is.null(max_n)) tab <- utils::head(tab, max_n)

  memb <- t(vapply(names(tab), function(k) {
    as.integer(strsplit(k, "", fixed = TRUE)[[1]]) == 1L
  }, logical(ncol(m))))
  dimnames(memb) <- list(NULL, colnames(m))

  list(size = as.integer(tab), membership = memb, sets = colnames(m),
       set_sizes = colSums(m), total = nrow(m))
}

#' Shiny bindings for upset
#'
#' Output and render functions for using [upset()] within Shiny applications and
#' interactive R Markdown / Quarto documents.
#'
#' @param output_id Output variable to read from.
#' @param width,height Element size, passed to
#'   [htmlwidgets::shinyWidgetOutput()].
#' @param expr An expression that generates an [upset()] widget.
#' @param env The environment in which to evaluate `expr`.
#' @param quoted Is `expr` already quoted? Defaults to `FALSE`.
#' @return `upsetOutput()` returns a Shiny output UI element; `renderUpset()`
#'   returns a Shiny render function.
#' @name upset-shiny
#' @export
upsetOutput <- function(output_id, width = "100%", height = "520px") {
  htmlwidgets::shinyWidgetOutput(output_id, "upset", width, height,
    package = "plotomics"
  )
}

#' @rdname upset-shiny
#' @export
renderUpset <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) {
    expr <- substitute(expr)
  }
  htmlwidgets::shinyRenderWidget(expr, upsetOutput, env, quoted = TRUE)
}
